Add support for linuxX64

ToxicBakery opened the pull request:
Adding support for Linux X86 as part of attempting to add support  in `apollo-kotlin`.

To locally run `check` without error I also needed to run `apiDump` though I was unsure if I should be committing those changes directly. I can revert the api changes if that was a mistake.

Ref: https://github.com/apollographql/apollo-kotlin/issues/3483

apollo-cla commented:
@ToxicBakery: Thank you for submitting a pull request! Before we can merge it, you'll need to sign the Apollo Contributor License Agreement here: https://contribute.apollographql.com/

BoD commented:
Hey thanks @ToxicBakery! 

It looks like you updated the Gradle build for `apollo-kotlin-mockserver` but what you want is actually the one for `apollo-runtime`. To do that you can pass `withLinux = true` instead of false. The same is true for `apollo-mpp-utils` (which is a dependency of `apollo-runtime`). For that one you'll need to implement a linux version of `currentTimeMillis()`.

The changes to mockserver should actually happen on the [dedicated repo](https://github.com/apollographql/apollo-kotlin-mockserver).

And you're correct about `apiDump` - it should be run and the changes be part of the PR.

Don't hesitate if you need guidance!

ToxicBakery commented:
Hey @BoD,

I've done that in the `apollo-runtime` module but that leads to a dependency chain of `apollo-mpp-utils` needing linux support as you have noted and I have that resolved by simply using kotlinx.date time, already a dependency in the repo, to provide the system clock millis. The issue is that for tests to pass the project depends on this repo for the mock server and so now I'm here adding support. Maybe I'm misunderstanding a step, this is where I'm at trying to run checks in the main apollo-kotlin repo:

```
Could not determine the dependencies of task ':apollo-runtime:linkDebugTestLinux'.
> Could not resolve all dependencies for configuration ':apollo-runtime:linuxTestCompileKlibraries'.
   > Could not resolve com.apollographql.mockserver:apollo-mockserver:0.1.0.
     Required by:
         project :apollo-runtime
      > No matching variant of com.apollographql.mockserver:apollo-mockserver:0.1.0 was found. The consumer was configured to find a library for use during 'kotlin-api', preferably optimized for non-jvm, as well as attribute 'org.jetbrains.kotlin.klib.packaging' with value 'non-packed', attribute 'org.jetbrains.kotlin.native.target' with value 'linux_x64', attribute 'org.jetbrains.kotlin.platform.type' with value 'native' but:
      ...
```

BoD commented:
I think we'll want to avoid making `apollo-mpp-utils` depend on `kotlinx-datetime`, as this would transitively add this dependency to `apollo-runtime` too. 

About the mockserver - I now realize that we'll need to first add the linux target there, and make a release, before we can add it here.

ToxicBakery commented:
I'll look into an alternative solution for the date time in mpp-utils. As for the mock server, I believe you may have been thrown off by the link to the apollo-kotlin repo, this PR is against the dedicated mockserver repo.

martinbonnin commented:
Thanks for looking into this! I can make a release for this. Before, we do that, can we add a specific ubuntu CI job to run linux tests? I believe the current job is running macos and not running the linux tests.

BoD commented:
> I believe you may have been thrown off by the link to the apollo-kotlin repo, this PR is against the dedicated mockserver repo.

Woops you're absolutely right sorry about this :) My comments were about the main repo.

ToxicBakery commented:
@martinbonnin are you asking me to add a Linux CI configuration or @BoD ?

BoD commented on apollo-mockserver/build.gradle.kts:
```diff
@@ -20,6 +20,7 @@ kotlin {
   tvosArm64()
   tvosX64()
   tvosSimulatorArm64()
+  linuxX64()
```
We should probably also add `linuxArm64()` while we're at it

ToxicBakery commented:
Added arm64 and also added the job for linux x86 testing. Github does not have Arm Linux instances that I'm aware of so I'm not sure if there is a reasonable way to test that arch.

martinbonnin commented:
Apollogies this slipped through. I just triggered the `0.1.1` release, should be live on Maven Central in a few minutes: https://github.com/apollographql/apollo-kotlin-mockserver/releases/tag/v0.1.1.
